prov/gni/src/gnix_vc.c

```diff
@@ -69,6 +70,367 @@ static int __gnix_vc_conn_ack_comp_fn(void *data);
 /*******************************************************************************
  * Helper functions
  ******************************************************************************/
+/**
+ * Insert vc based on AV type and favor FI_AV_TABLE
+ *
+ * @assumption: ep, ep->av, index, key are all valid non-null ptrs.
+ */
+static inline int __gnix_vc_insert(struct gnix_fid_ep *ep,
+				   fi_addr_t *index,
+				   gnix_ht_key_t *key,
+				   struct gnix_vc *vc)
+{
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+	int ret, av_type = ep->av->type;
+
+	if (likely(av_type == FI_AV_TABLE)) {
+		fastlock_acquire(&ep->vc_lock);
+		ret = _gnix_vec_insert_at(ep->vc_table, (void *) vc, *index);
+
+		if (ret == -FI_ENOSPC) {
+			fastlock_release(&ep->vc_lock);
+			return ret;
+		}
+
+
+		if (unlikely(ret != FI_SUCCESS)) {
+			GNIX_FATAL(FI_LOG_EP_CTRL,
+				   "_gnix_vec_insert_at returned %s\n",
+				   fi_strerror(-ret));
+		}
+
+		vc->modes |= GNIX_VC_MODE_IN_TABLE;
+		fastlock_release(&ep->vc_lock);
+	} else if (av_type == FI_AV_MAP){
+		fastlock_acquire(&ep->vc_lock);
+		ret = _gnix_ht_insert(ep->vc_ht, *key, (void *) vc);
+
+		/* vc was inserted into ht between call to lookup and here! */
+		if (ret == -FI_ENOSPC) {
+			fastlock_release(&ep->vc_lock);
+			return ret;
+		}
+
+		if (unlikely(ret != FI_SUCCESS)) {
+			GNIX_FATAL(FI_LOG_EP_CTRL,
+				   "_gnix_ht_insert returned %s\n",
+				   fi_strerror(-ret));
+		}
+
+		vc->modes |= GNIX_VC_MODE_IN_HT;
+		fastlock_release(&ep->vc_lock);
+	} else {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid AV type \"%d\" in __gnix_vc_insert\n",
+			  av_type);
+                return -FI_EINVAL;
+	}
+
+	return ret;
+}
+
+/**
+ * Lookup VC based on AV type and favor FI_AV_TABLE.  If the AV type
+ * is FI_AV_TABLE, then table_dest_addr will be populated. If the AV type
+ * is FI_AV_MAP, then key and ht_dest_addr will be populated.  In either
+ * case av_entry and vc_ptr will be populated.
+ *
+ * @assumption: The user correctly sets is_dest_addr_gnix_addr;
+ * if key is non-null then it is the correct key for the given dest_addr;
+ * ep is non-null;
+ * vc_ptr is non-null.
+ *
+ * TODO: Only lock the vc_lock when looking up the VC.
+ */
+static inline int __gnix_vc_lookup_vc(struct gnix_fid_ep *ep, void *dest_addr,
+				      bool is_dest_addr_gnix_addr,
+				      struct gnix_av_addr_entry **av_entry,
+				      gnix_ht_key_t **key,
+				      fi_addr_t *table_dest_addr,
+				      struct gnix_address *ht_dest_addr,
+				      struct gnix_vc **vc_ptr)
```

Comment:
Would a static variable for the hashtable key be acceptable?  So, rather than mallocing the key we could add “static gnix_ht_key_t *key_ptr” to __gnix_vc_lookup_vc and set the parameter key to &key_ptr.
> On Mar 28, 2016, at 10:16 AM, Howard Pritchard <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_vc.c <https://github.com/ofi-cray/libfabric-cray/pull/736#discussion_r57590472>:
> 
> > + * case av_entry and vc_ptr will be populated.
> > + *
> > + * @assumption: The user correctly sets is_dest_addr_gnix_addr;
> > + * if key is non-null then it is the correct key for the given dest_addr;
> > + * ep is non-null;
> > + * vc_ptr is non-null.
> > + *
> > + * TODO: Only lock the vc_lock when looking up the VC.
> > + */
> > +static inline int __gnix_vc_lookup_vc(struct gnix_fid_ep *ep, void *dest_addr,
> > +				      bool is_dest_addr_gnix_addr,
> > +				      struct gnix_av_addr_entry **av_entry,
> > +				      gnix_ht_key_t **key,
> > +				      fi_addr_t *table_dest_addr,
> > +				      struct gnix_address *ht_dest_addr,
> > +				      struct gnix_vc **vc_ptr)
> I think this function is going to need to be refactored some. The problem is this function is in the critical path, so its essential that malloc's not been done there. This may need to be split up in to a slower version that returns the key and one that doesn't.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/ofi-cray/libfabric-cray/pull/736/files/408e2a2ddff4bc9751170dc9ac6f6a66a970679b#r57590472>